csharp/src/Ice/OutputStream.cs

```diff
@@ -1164,21 +1167,40 @@ private static int Distance(IList<ArraySegment<byte>> data, Position start, Posi
             }
         }
 
-        /// <summary>Writes a size into a span of bytes using a fixed number of bytes (currently always 4 bytes).
-        /// </summary>
+        /// <summary>Writes a size into a span of bytes using a fixed number of bytes.</summary>
         /// <param name="size">The size to write.</param>
-        /// <param name="data">The destination byte buffer.</param>
+        /// <param name="data">The destination byte buffer, which must be 1, 2 or 4 bytes long.</param>
         private static void WriteFixedLength20Size(int size, Span<byte> data)
         {
+            int sizeLength = data.Length;
+            Debug.Assert(sizeLength == 1 || sizeLength == 2 || sizeLength == 4);
+
             if (size < 0 || size > 1_073_741_823) // 2^30 -1
             {
                 throw new ArgumentOutOfRangeException("size is out of range", nameof(size));
             }
 
+            Span<byte> uintBuf = stackalloc byte[4];
             uint v = (uint)size;
             v <<= 2;
-            v |= (uint)0x02;
-            MemoryMarshal.Write(data, ref v);
+
+            uint encodedLength = sizeLength switch
+            {
+                1 => 0x00,
+                2 => 0x01,
+                _ => 0x02
+            };
+            v |= encodedLength;
+            MemoryMarshal.Write(uintBuf, ref v);
+            for (int i = sizeLength; i < 4; ++i)
+            {
+                if (uintBuf[i] != 0)
+                {
+                    throw new ArgumentOutOfRangeException($"size `{size}' does not fit in {sizeLength} bytes",
+                        nameof(size));
+                }
+            }
+            uintBuf.Slice(0, sizeLength).CopyTo(data);
```

Comment:
what about using `checked` for the range checks

```
Debug.Assert(data.Length == 1 || data.Length == 2 || data.Length == 4);
            checked
            {
                // Left shift using checked multiplication to ensure there is not overflow
                uint v = (uint)size * 4;
                switch (data.Length)
                {
                    case 1:
                    {
                        data[0] = (byte)v;
                        break;
                    }
                    case 2:
                    {
                        ushort value = (ushort)v;
                        value |= 0x1;
                        MemoryMarshal.Write(data, ref value);
                        break;
                    }
                    default:
                    {
                        v |= 0x2;
                        MemoryMarshal.Write(data, ref v);
                        break;
                    }
                }
            }
```

Reply:
I considered, but this would only remove half this check:

```
            if (size < 0 || size > 1_073_741_823) // 2^30 -1
            {
                throw new ArgumentOutOfRangeException("size is out of range", nameof(size));
            }
```

`WriteVarULong` and `WriteVarLong` don't use checked arithmetic either.